src/Open3D/Visualization/Rendering/Filament/FilamentResourceManager.cpp

```diff
@@ -0,0 +1,718 @@
+// use engine deleter for allocated filament entities
+template <class ResourceType>
+std::shared_ptr<ResourceType> MakeShared(ResourceType* pointer,
+                                         filament::Engine& engine) {
+    return std::move(std::shared_ptr<ResourceType>(
+            pointer, [&engine](ResourceType* p) { engine.destroy(p); }));
+}
+
+template <class Handle, class ResourceType>
+Handle RegisterResource(filament::Engine& engine,
+                        ResourceType* resource,
+                        ResourcesContainer<ResourceType>& container) {
+    if (!resource) {
+        utility::LogError("Trying to register empty resource!");
+        return Handle::kBad;
+    }
+
+    auto newHandle = Handle::Next();
+    container[newHandle] = std::move(MakeShared(resource, engine));
+
+    return newHandle;
+}
+
+template <class ResourceType>
+std::weak_ptr<ResourceType> FindResource(
+        const REHandle_abstract& id,
+        ResourcesContainer<ResourceType>& container) {
+    auto found = container.find(id);
+    if (found != container.end()) {
+        return found->second;
+    }
+
+    utility::LogWarning("Resource {} not found.", id);
+    return std::weak_ptr<ResourceType>();
+}
+
+template <class ResourceType>
+void DestroyResource(const REHandle_abstract& id,
+                     ResourcesContainer<ResourceType>& container) {
+    auto found = container.find(id);
+    if (found == container.end()) {
+        utility::LogError("Trying to destroy nonexistent resource ({})!", id);
+        return;
+    }
+
+    container.erase(found);
+}
+
+// Image data that retained by renderer thread,
+// will be freed on PixelBufferDescriptor callback
+std::unordered_map<std::uint32_t, std::shared_ptr<geometry::Image>>
+        pendingImages;
+
+std::intptr_t RetainImageForLoading(
+        const std::shared_ptr<geometry::Image>& img) {
+    static std::intptr_t imgId = 1;
+
+    const auto id = imgId;
+    pendingImages[imgId] = img;
+    ++imgId;
+
+    return id;
+}
+
+void FreeRetainedImage(void* buffer, size_t size, void* userPtr) {
+    const auto id = reinterpret_cast<std::intptr_t>(userPtr);
+    auto found = pendingImages.find(id);
+    if (found != pendingImages.end()) {
+        pendingImages.erase(found);
+    } else {
+        utility::LogDebug(
+                "Trying to release non existent image shared pointer, id: {}",
+                id);
+    }
+}
+
+filament::Material* LoadMaterialFromFile(const std::string& path,
+                                         filament::Engine& engine) {
+    std::vector<char> materialData;
+    std::string errorStr;
+
+    if (utility::filesystem::FReadToBuffer(path, materialData, &errorStr)) {
+        using namespace filament;
+        return Material::Builder()
+                .package(materialData.data(), materialData.size())
+                .build(engine);
+    }
+
+    utility::LogDebug("Failed to load default material from {}. Error: {}",
```

Comment:
I think we need to figure out how we want to report errors while loading a model. We can't just blindly put up a popup message, since we are likely to have multiple failures from a model if we have one, and it's never fun clicking through 20 alerts. I've made a task in Asana to deal with errors and asserts.